FormalSystems/Computability/Loop.lean

```diff
@@ -0,0 +1,170 @@
+import Mathlib.Data.Finite.Defs
+import Mathlib.Data.Finset.Fold
+import Mathlib.Data.Finset.Prod
+
+import FormalSystems.Preliminaries.Fold
+
+structure Variable where
+  index: Nat
+  deriving DecidableEq
+
+theorem Variable.eq (v₁ v₂: Variable): v₁ = v₂ ↔ v₁.index = v₂.index := by
+  constructor
+  . intro; congr
+  . intro h
+    cases v₁; cases v₂
+    dsimp at h; rw [h]
+
+inductive LoopProgram
+| add (x y: Variable) (n: Nat)
+| sub (x y: Variable) (n: Nat)
+| concat (p₁ p₂: LoopProgram)
+| loop (x: Variable) (p: LoopProgram)
+deriving DecidableEq
+
+def LoopProgram.toFunction (p: LoopProgram):
+  Nat → Nat :=
+  -- Todo: define how to evaluate `p`, with x₀ bound to the input
+  -- for example one could use a state monad with an `AssocList`
+  sorry
+
+def LoopProgram.len: LoopProgram → Nat
+| add x y n => x.index + y.index + n + 1
+| sub x y n => x.index + y.index + n + 1
+| concat p₁ p₂ => p₁.len + p₂.len + 1
+| loop n p => p.len + n.index + 1
```

Comment:
Intuitively, I would define `len` as follows but maybe you have something different in mind which I do not understand. In particular, it is not clear to me why the length of the programs depends on variables and numbers that occur in the program.
```suggestion
def LoopProgram.len: LoopProgram → Nat
| add x y n => 1
| sub x y n => 1
| concat p₁ p₂ => p₁.len + p₂.len
| loop n p => p.len + 1
```

Reply:
> In particular, it is not clear to me why the length of the programs depends on variables and numbers that occur in the program.

The length is used to show to that the Loop busy-beaver (mapping `n` to the largest output a Loop program of length `n` can generate) is not Loop-computable, but computable. If the length is independent of constants, this is not well-defined. See also [TheoLog, lecture 3](https://iccl.inf.tu-dresden.de/w/images/d/d5/TheoLog2021-Vorlesung-03-overlay.pdf).

Reply:
I see; it is somehow clear that this is required for the proof. I was still a bit thrown off since e.g. `42` and `73` have the same length (intuitively).
But probably this also works with a unary encoding as it is essentially used here(?)

Reply:
Yeah, this essentially uses a unary encoding for numbers to derive the length, which is more than is required for the proof, but still works (and might be easier to work with than, e.g., the decimal encoding the lecture uses).